### PR TITLE
Baremetal support in bosh-softlayer-cpi

### PR DESCRIPTION
**Pull request by jianqiu**
This PR is based on PR [126](https://github.com/cloudfoundry/bosh-softlayer-cpi/pull/126), I had refined the some methods to make it clean.

### CODE REVIEW

**Comment by jianqiu** [comment]
@maximilien 
Could you please look into the codes as soon as possible, because this PR contains lots of changes, to avoid more conflicts in future, could you please review it as high priority? Thanks!

Hi, @sykesm,  could you please help us review the codes, we are also looking forward to your comments, thanks!

**Comment by maximilien** [comment]
Hi, @jianqiu, unfortunately you need to rebase. This is due to my merging @mattcui PR that was submitted before yours. Sorry about that.

Also, please squash your commits into one. Thanks and I will review ASAP.

**Comment by jianqiu** [comment]
Hi, @maximilien All conflicts had been resolved, but I did not find a good solution to squash commits into one, I had tried many times, unfortunately no one succeeded. Do you have a chance to squash them when merging? Thanks!

**Comment by jianqiu** [comment]
@maximilien @mattcui 
Could you please review the PR as soon as possible? There is another PR pending on it. Thanks!

File: action/attach_disk.go [hunk 1]
```diff
@@ -7,19 +7,21 @@ import (
...
 	vmFinder   bslcvm.Finder
 	diskFinder bslcdisk.Finder
 }
 
-func NewAttachDisk(vmFinder bslcvm.Finder, diskFinder bslcdisk.Finder) AttachDisk {
-	return AttachDisk{
-		vmFinder:   vmFinder,
-		diskFinder: diskFinder,
-	}
+func NewAttachDisk(
```
Comment (maximilien) [inline action/attach_disk.go]: If you are going to change the name to include `Action` then why not also rename the `New...` function to also include the `Action` suffix? Consistency is important. Also, are all the other also named with `Action`?

Comment (jianqiu) [inline action/attach_disk.go]: Yes, all the other are named with "Action" suffix. What about submit another PR to cover your comment?  Since it is a little hard to combine them into it. Thanks!

**Comment by maximilien** [comment]
BTW, @jianqiu, reason it's taking so long for me to review this is that the changes are so much at once that loading the diff crashed my chrome twice, I may need to use s more powerful machine... I'll attempt again to finish this review today

**Comment by jianqiu** [comment]
@maximilien 
Thanks for you great efforts. It really has changed a lot, I knew it was not good, we will submit more small er PR in future.
BTW, after the PR merged, we will consider how to merge the 'portable ip' support suggested by @sykesm

File: action/concrete_factory_test.go [hunk 2]
```diff
@@ -7,24 +7,11 @@ import (
...
-	fakesys "github.com/cloudfoundry/bosh-utils/system/fakes"
-
-	fakeslclient "github.com/maximilien/softlayer-go/client/fakes"
-
-	bslcdisk "github.com/cloudfoundry/bosh-softlayer-cpi/softlayer/disk"
-	bslcstem "github.com/cloudfoundry/bosh-softlayer-cpi/softlayer/stemcell"
-	bslcvm "github.com/cloudfoundry/bosh-softlayer-cpi/softlayer/vm"
 )
 
 var _ = Describe("concreteFactory", func() {
```
Comment (maximilien) [inline action/concrete_factory_test.go]: This should be `Describe("ConcreteFactory", ...)`

File: action/has_vm_test.go [hunk 3]
```diff
@@ -23,16 +23,10 @@ var _ = Describe("HasVM", func() {
 	})
 
 	Describe("Run", func() {
-		It("tries to find VM with given VM CID", func() {
-			_, err := action.Run(1234)
-			Expect(err).ToNot(HaveOccurred())
```
Comment (maximilien) [inline action/has_vm_test.go]: Where is this test now?

File: config/config_test.go [hunk 4]
```diff
@@ -59,7 +59,7 @@ var _ = Describe("NewConfigFromPath", func() {
 		err := fs.WriteFileString("/config.json", "{}")
 		Expect(err).ToNot(HaveOccurred())
 
-		_, err = NewConfigFromPath("/config.json", fs)
+		_, err = config.NewConfigFromPath("/config.json", fs)
```
Comment (maximilien) [inline config/config_test.go]: We should create temp dirs and files and clean up after

**Comment by maximilien** [comment]
@jianqiu merging but this is the last time I will accept a PR that has so many unrelated changes. From now on, submit smaller PRs more often. Thx
